ouroboros/src/validator/mod.rs

```diff
@@ -1,6 +1,33 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum ValidationError {
+    #[error("InvalidByteLength: {0}")]
+    InvalidByteLength(String),
+    #[error("Invalid VRF key for pool: expected: {0}, was: {1}")]
+    InvalidVrfKeyForPool(String, String),
+    #[error("Invalid block hash, expected: {0}, was: {1}")]
+    InvalidBlockHash(String, String),
+    #[error("Ledger error: {0}")]
+    LedgerError(#[from] crate::ledger::Error),
+    #[error("VrfVerificationError: {0}")]
+    VrfVerificationError(#[from] pallas_crypto::vrf::VerificationError),
+    #[error("InvalidVrfProofHash, expected: {0}, was: {1}")]
+    InvalidVrfProofHash(String, String),
+    #[error("InvalidVrfLeaderHash, expected: {0}, was: {1}")]
+    InvalidVrfLeaderHash(String, String),
+    #[error("InvalidOpcertSequenceNumber: {0}")]
+    InvalidOpcertSequenceNumber(String),
+    #[error("InvalidOpcertSignature")]
+    InvalidOpcertSignature,
+    #[error("KesVerificationError: {0}")]
+    KesVerificationError(String),
+    #[error("InsufficientPoolStake")]
+    InsufficientPoolStake,
```

Comment:
It doesn't feel right to have hex-encoded strings here as part of the `ValidationError` types for many of those fields. It is IMO a mismatch of responsibility / representation. Having an hex-encoded string is useful for display and should be the responsibility of the `Debug` or `Display` trait's implementation. 

To give an concrete example where it can be problematic; if you end up returning the error and handling it from another place, then you end up paying the cost of hex-encoding the data for nothing. It wouldn't be problematic in a lazy language, but in Rust that'd be wasted CPU.

Reply:
I would decouple the representation of the error from its definition: leave raw types in the definition and provide different representations elsewhere (a `Display` trait is one option, but there's also the case of rendering errors as JSON or any other representation as part of structured logging)